docs(blog): Jest 27 blog post (#11131)
facebook/jest: website/blog/2021-03-11-jest-27.md

> ---

Author: Gotta remember to update date

Reviewer: tuesday 25th is the plan if you wanna prep it

Author: I put that for now :+1:

title: 'Jest 27: New Defaults for Jest, 2021 edition ⏩'
author: Tim Seckinger
authorURL: https://github.com/jeysal
authorImageURL: https://avatars.githubusercontent.com/u/16069751
---

> In the [Jest 26 blog post](/blog/2020/05/05/jest-26) almost a year ago, we announced that after two major releases with few breaking changes, Jest 27 will flip some switches to set better defaults for projects that are new or can migrate smoothly. This gives us the opportunity to remove some packages from the default distribution of Jest 28 and publish them as separately installable and pluggable modules instead. Everyone on the new defaults can benefit from a smaller install size, while people needing these packages can still install them separately.

Reviewer: ```suggestion
In the [Jest 26 blog post](/blog/2020/05/05/jest-26) more than a year ago, we announced that after two major releases with few breaking changes, Jest 27 will flip some switches to set better defaults for projects that are new or can migrate smoothly. This gives us the opportunity to remove some packages from the default distribution of Jest 28 and publish them as separately installable and pluggable modules instead. Everyone on the new defaults can benefit from a smaller install size, while people needing these packages can still install them separately.
```

😅

Author: oof

Author: Let's do `about` :D

The main reason why it took us so long to land this was, somewhat surprisingly, an out of memory error on our build pipeline. It turns out that the dependencies we load for each test file to perform the parsing, snapshot insertion, and printing do incur a significant time and memory overhead.  
So with some [tricks](https://github.com/facebook/jest/issues/9898), we've **sped up the initialization per test file** by roughly 70% compared to Jest 26. Note that you will almost certainly not see this big of a performance improvement on your project—you would need a lot of test files that each run very quickly to best notice this, and the overhead when using a [JSDOM environment](/docs/configuration#testenvironment-string) dwarfs any such improvement.

> In other news, the [native ESM support](https://github.com/facebook/jest/issues/9430) is progressing, but some major complexities, for instance around mocking, are still ahead of us, and we continue to observe the migration to ESM as a huge ecosystem effort, where Node and a lot of crucial tools and packages all have to rely on each other to deliver an overall compelling experience.

Reviewer: maybe mention something about #11167? It's not for the tests themselves, but the stuff around it

Author: done

After years of large-scale use at Facebook and of course in Jest itself, as well as recent adoption in `create-react-app`, we are now confident that `jest-circus` is highly compatible with `jest-jasmine2` and should work in most environments with little to no migration work. There may be minor differences in execution order and strictness, but we expect no major upgrade difficulties other than for code relying on Jasmine-specific APIs such as `jasmine.getEnv()`. If you rely extensively on such APIs, you can opt back in to the Jasmine-based test runner by [configuring](/docs/configuration#testrunner-string) `"testRunner": "jest-jasmine2"`.

Running tests in a [JSDOM environment](/docs/configuration#testenvironment-string) incurs a significant performance overhead. Because this was the default behavior of Jest unless otherwise configured up until now, users who are writing Node apps, for example, may not even know they are given an expensive DOM environment that they do not even need.  

> For this reason, we are **changing the default test environment** from `"jsdom"` to `"node"`. If you are affected by this change because you use DOM APIs and do not have the test environment explicitly configured, you should be receiving an error when e.g. `document` is accessed, and you can configure `"testEnvironment": "jsdom"` or use [per-file environment configuration](/docs/configuration#testenvironment-string) to resolve this.  

Reviewer: ```suggestion
For this reason, we are **changing the default test environment** from `"jsdom"` to `"node"`. If you are affected by this change because you use DOM APIs and do not have the test environment explicitly configured, you should be receiving an error when e.g. `document` is accessed, and you can configure `"testEnvironment": "jsdom"` or use [per-file environment configuration](Configuration.md#testenvironment-string) to resolve this.  
```

does this work from blog posts? I forget (and feel like I've asked before)

Author: Last time it didn't even with docu v2, did any infra change since then? :sweat_smile:

Reviewer: nope 😀 I'm just getting old and memory is going 😛

Author: Tried once more just now, nope, still the same :sweat_smile:

For mixed projects where some tests require a DOM environment but others don't, we recommend using the fast `"node"` environment by default and declaring exactly those tests that need the DOM using [docblocks](/docs/configuration#testenvironment-string).  
In the next major, we plan to also eliminate `jest-jasmine2` and `jest-environment-jsdom` from the Jest dependency tree and require them to be installed explicitly, so that many users can profit from a smaller install size with less clutter that they don't need.

Another default that we are changing affects Fake Timers aka [Timer Mocks](/docs/timer-mocks). We introduced an opt-in "modern" implementation of Fake Timers in Jest 26 accessed transparently through the same API, but with much more comprehensive mocking, such as for `Date` and `queueMicrotask`.  

> This **modern fake timers implementation will now be the default**. If you are among the unlucky few who are affected by the subtle implementation differences too heavily to migrate, you can get back the old implementation using `jest.useFakeTimers("legacy")` or, if you are enabling fake timers globally via [configuration](/docs/configuration#timers-string), `"timers": "legacy"`.

Reviewer: ```suggestion
This **modern fake timers implementation will now be the default**. If you are among the unlucky few who are affected by the subtle implementation differences too heavily to migrate, you can get back the old implementation using `jest.useFakeTimers("legacy")` or, if you are enabling fake timers globally via [configuration](Configuration.md#timers-string), `"timers": "legacy"`.
```

## Features coming with breaking changes

We introduced a few more small breaking changes to help you avoid mistakes by disallowing some things that can easily happen unintentionally:

- The same `done` test callback may not be called more than once,
- calling `done` and returning a Promise may not be combined,
- a `describe` block must not return anything,

and we [made some TypeScript types stricter](https://github.com/facebook/jest/pull/10512).

Modules used in the following configuration options are now transformed like the rest of your code, which may be breaking if you relied on them being loaded as-is:

- `testEnvironment`
- `runner`
- `testRunner`

> - `snapshotResolver` <!-- TODO pending merge https://github.com/facebook/jest/pull/8829 -->

Reviewer: this has been merged

## Miscellaneous breaking changes

We removed some long-deprecated functions:

- `jest.addMatchers` (use `expect.extend` instead)
- `jest.resetModuleRegistry` (use `jest.resetModules` instead)
- `jest.runTimersToTime` (use `jest.advanceTimersByTime` instead)

A lot of Jest's packages have been migrated to use ESM-style exports (although they are still shipped as CommonJS), so if you consume e.g. `pretty-format` directly, you may need to adjust your import to a `default` import.

> We dropped support for Node 13 — but Jest always supports the _Current_ and all _LTS_ [Node versions](https://nodejs.org/en/about/releases/).

Reviewer: A note we're keeping support for node 10 until v28, which will _also_ drop v15?

The main reason why it took us so long to land this was, somewhat surprisingly, an out of memory error on our build pipeline. It turns out that the dependencies we load for each test file to perform the parsing, snapshot insertion, and printing do incur a significant time and memory overhead.  
So with some [tricks](https://github.com/facebook/jest/issues/9898), we've **sped up the initialization per test file** by roughly 70% compared to Jest 26. Note that you will almost certainly not see this big of a performance improvement on your project—you would need a lot of test files that each run very quickly to best notice this, and the overhead when using a [JSDOM environment](/docs/configuration#testenvironment-string) dwarfs any such improvement.

In other news, the [native ESM support](https://github.com/facebook/jest/issues/9430) is progressing, but some major complexities, for instance around mocking, are still ahead of us, and we continue to observe the migration to ESM as a huge ecosystem effort, where Node and a lot of crucial tools and packages all have to rely on each other to deliver an overall compelling experience.

> <!-- TODO pending merge https://github.com/facebook/jest/pull/9351 -->

Reviewer: merged

So with some [tricks](https://github.com/facebook/jest/issues/9898), we've **sped up the initialization per test file** by roughly 70% compared to Jest 26. Note that you will almost certainly not see this big of a performance improvement on your project—you would need a lot of test files that each run very quickly to best notice this, and the overhead when using a [JSDOM environment](/docs/configuration#testenvironment-string) dwarfs any such improvement.

In other news, the [native ESM support](https://github.com/facebook/jest/issues/9430) is progressing, but some major complexities, for instance around mocking, are still ahead of us, and we continue to observe the migration to ESM as a huge ecosystem effort, where Node and a lot of crucial tools and packages all have to rely on each other to deliver an overall compelling experience.

<!-- TODO pending merge https://github.com/facebook/jest/pull/9351 -->

We've also merged [a PR](https://github.com/facebook/jest/pull/9351) to be able to deal with test files symlinked into the test directory, a feature much wanted by users of [Bazel](https://bazel.build/).

> <!-- TODO pending merge https://github.com/facebook/jest/issues/9504 -->

Reviewer: merged

## Features coming with breaking changes

We introduced a few more small breaking changes to help you avoid mistakes by disallowing some things that can easily happen unintentionally:

- The same `done` test callback may not be called more than once,
- calling `done` and returning a Promise may not be combined,
- a `describe` block must not return anything,

and we [made some TypeScript types stricter](https://github.com/facebook/jest/pull/10512).

Modules used in the following configuration options are now transformed like the rest of your code, which may be breaking if you relied on them being loaded as-is:

- `testEnvironment`
- `runner`
- `testRunner`
- `snapshotResolver` <!-- TODO pending merge https://github.com/facebook/jest/pull/8829 -->

## Miscellaneous breaking changes

We removed some long-deprecated functions:

- `jest.addMatchers` (use `expect.extend` instead)
- `jest.resetModuleRegistry` (use `jest.resetModules` instead)
- `jest.runTimersToTime` (use `jest.advanceTimersByTime` instead)

> A lot of Jest's packages have been migrated to use ESM exports, so if you consume e.g. `pretty-format` directly, you may need to adjust your import to a `default` import.

Reviewer: Should probably mentions it's still CJS via Babel's transform. It's not _native_ ESM
